File: app/src/main/java/com/app/fortuneapp/activities/MainActivity.java
package com.app.fortuneapp.activities;

import static com.app.fortuneapp.chat.constants.IConstants.CATEGORY;
import static com.app.fortuneapp.chat.constants.IConstants.EXTRA_USER_ID;
import static com.app.fortuneapp.chat.constants.IConstants.NAME;
import static com.app.fortuneapp.chat.constants.IConstants.TICKET_ID;
import static com.app.fortuneapp.chat.constants.IConstants.TYPE;
import static com.app.fortuneapp.helper.Constant.DESCRIPTION;
import static com.app.fortuneapp.helper.Constant.REFERRED_BY;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.FragmentManager;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.app.fortuneapp.R;
import com.app.fortuneapp.chat.MessageActivity;
import com.app.fortuneapp.chat.TicketFragment;
import com.app.fortuneapp.chat.models.Ticket;
import com.app.fortuneapp.fragment.FaqFragment;
import com.app.fortuneapp.fragment.FindMissingFragment;
import com.app.fortuneapp.fragment.HomeFragment;
import com.app.fortuneapp.fragment.InFoFragment;
import com.app.fortuneapp.fragment.ProfileFragment;
import com.app.fortuneapp.fragment.WalletFragment;
import com.app.fortuneapp.helper.ApiConfig;
import com.app.fortuneapp.helper.Constant;
import com.app.fortuneapp.helper.DatabaseHelper;
import com.app.fortuneapp.helper.Session;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.messaging.FirebaseMessaging;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;


public class MainActivity extends AppCompatActivity {

    public static FragmentManager fm = null;
    private BottomNavigationView navbar;
    Activity activity;
    Session session;
    String NOTIFY_CHAT;
    Dialog dialog = null;
    long fetch_time;
    DatabaseHelper databaseHelper;
    String RandomId;
    DatabaseReference reference;
    Button btnChatUs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fm = getSupportFragmentManager();
        setContentView(R.layout.activity_main);
        navbar = findViewById(R.id.bottomNavigation);
        navbar.setSelectedItemId(R.id.Home);
        activity = MainActivity.this;
        databaseHelper = new DatabaseHelper(activity);
        session = new Session(activity);
        NOTIFY_CHAT = getIntent().getStringExtra("NOTIFY_CHAT");
        if (session.getBoolean(Constant.CHECK_NOTIFICATION)) {
            showReadNotificationPopup();
        }

        //importUrl();


        if (NOTIFY_CHAT != null) {
            if (NOTIFY_CHAT.equals("join_chat")) {
                checkJoining();
            }else {
                navbar.setSelectedItemId(R.id.Support);
                fm.beginTransaction().replace(R.id.Container, new TicketFragment()).commit();


            }


        } else {
            if (session.getData(Constant.STATUS).equals("0")) {
                fm.beginTransaction().replace(R.id.Container, new InFoFragment()).commitAllowingStateLoss();
            } else {
                if (session.getData(Constant.TASK_TYPE).equals("champion")) {
                    fm.beginTransaction().replace(R.id.Container, new FindMissingFragment()).commit();
                } else {
                    fm.beginTransaction().replace(R.id.Container, new HomeFragment()).commit();
                }
            }

        }

        navbar.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.Profile:
                        fm.beginTransaction().replace(R.id.Container, new ProfileFragment()).commitAllowingStateLoss();
                        break;
                    case R.id.Home:
                        if (session.getData(Constant.STATUS).equals("0")) {
                            fm.beginTransaction().replace(R.id.Container, new InFoFragment()).commitAllowingStateLoss();
                        } else {
                            if (session.getData(Constant.TASK_TYPE).equals("champion"))
                                fm.beginTransaction().replace(R.id.Container, new FindMissingFragment()).commitAllowingStateLoss();
                            else
                                fm.beginTransaction().replace(R.id.Container, new HomeFragment()).commitAllowingStateLoss();
                        }
                        break;
                    case R.id.Wallet:
                        try {
                            fetch_time = Long.parseLong(session.getData(Constant.FETCH_TIME)) * 1000;
                        } catch (Exception e) {
                            fetch_time = 5 * 1000;


                        }
                        showWallet();

                        break;

                    case R.id.Support:
                        if (session.getData(Constant.STATUS).equals("0")) {
                            checkJoining();
                        } else {
                            fm.beginTransaction().replace(R.id.Container, new FaqFragment()).commitAllowingStateLoss();
                        }
                        break;
                }
                return true;
            }
        });
        FirebaseMessaging.getInstance().getToken().addOnSuccessListener(token -> {
            session.setData(Constant.FCM_ID, token);
        });

    }
    private void checkJoining() {
        //DatabaseReference ref = FirebaseDatabase.getInstance().getReference(Constant.JOINING_TICKET).child(session.getData(Constant.MOBILE));
        FirebaseDatabase.getInstance()
                .getReference(Constant.JOINING_TICKET).child(session.getData(Constant.MOBILE)).addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        if (dataSnapshot.exists()) {
                            Ticket user = dataSnapshot.getValue(Ticket.class);
                            Log.d("NOT_EXIST",user.getType() + " - "+session.getData(Constant.MOBILE));

                            sendChat(user.getId(),user.getName(),user.getCategory(),user.getType(),user.getDescription());

                        }
                        else {
                            joinTicket();

                        }
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });
    }

    private void joinTicket() {
        Long tsLong = System.currentTimeMillis()/1000;
        RandomId = session.getData(Constant.USER_ID) +"_"+ tsLong.toString();
        reference = FirebaseDatabase.getInstance().getReference(Constant.JOINING_TICKET).child(session.getData(Constant.MOBILE));
        HashMap<String, Object> hashMap = new HashMap<>();
        hashMap.put(Constant.ID, RandomId);
        hashMap.put(Constant.CATEGORY, "Joining");
        hashMap.put(Constant.DESCRIPTION, "Enquiry For Joining");
        hashMap.put(Constant.USER_ID, session.getData(Constant.USER_ID));
        hashMap.put(Constant.NAME, session.getData(Constant.NAME));
        hashMap.put(Constant.MOBILE, session.getData(Constant.MOBILE));
        hashMap.put(Constant.TYPE, Constant.JOINING_TICKET);
        hashMap.put(Constant.SUPPORT, "Admin");
        hashMap.put(Constant.REFERRED_BY, session.getData(REFERRED_BY));
        hashMap.put(Constant.TIMESTAMP, tsLong.toString());
        reference.setValue(hashMap).addOnCompleteListener(task1 -> {

            sendChat(RandomId, session.getData(Constant.NAME), "Joining", Constant.JOINING_TICKET, "Enquiry For Joining");

        });
    }

    private void sendChat(String id, String name, String category, String type, String description) {

        //Log.d("CHAT_DETAILS","USER_ID - "+id + "\nName - "+name+"\nCategory - "+category+"\nType - "+type +"Description - "+description);
        final Intent intent = new Intent(activity, MessageActivity.class);
        intent.putExtra(EXTRA_USER_ID, id);
        intent.putExtra(TICKET_ID, id);
        intent.putExtra(NAME, name);
        intent.putExtra(TYPE, type);
        intent.putExtra(DESCRIPTION, description);
        intent.putExtra(CATEGORY, category);
        startActivity(intent);
    }

    private void importUrl() {
        Map<String, String> params = new HashMap<>();
        ApiConfig.RequestToVolley((result, response) -> {
            if (result) {
                try {
                    JSONObject jsonObject = new JSONObject(response);
                    if (jsonObject.getBoolean(Constant.SUCCESS)) {
                        databaseHelper.deleteUrls();
                        JSONObject object = new JSONObject(response);
                        JSONArray jsonArray = object.getJSONArray(Constant.DATA);
                        for (int i = 0; i < jsonArray.length(); i++) {
                            JSONObject jsonObject1 = jsonArray.getJSONObject(i);
                            if (jsonObject1 != null) {
                                databaseHelper.AddtoUrl(jsonObject1.getString(Constant.ID), jsonObject1.getString(Constant.URL), "0");
                            } else {
                                break;
                            }
                        }


                    } else {


                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, activity, Constant.IMPORT_URLS, params, false);

    }

    private void showAlertDialogExplore() {


        // Create an alert builder
        AlertDialog.Builder builder
                = new AlertDialog.Builder(this);
        builder.setTitle("Explore");

// set the custom layout
        final View customLayout
                = getLayoutInflater()
                .inflate(
                        R.layout.explore_dialog,
                        null);
        builder.setView(customLayout);

// add a button
        builder.setPositiveButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(
                    DialogInterface dialog,
                    int which) {

// send data from the
//// AlertDialog to the Activity
            }
        });


        CardView cvSharechat = customLayout.findViewById(R.id.cvSharechat);
        CardView cvShorts = customLayout.findViewById(R.id.cvShorts);
        CardView cvGames = customLayout.findViewById(R.id.cvGames);
        AlertDialog dialog = builder.create();


        cvSharechat.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.dismiss();
                exploreApi("share_chat");
                Intent intent = new Intent(activity, LoadWebViewActivity.class);
                intent.putExtra(Constant.URL, "https://sharechat.com/");
                startActivity(intent);

            }
        });
        cvShorts.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.dismiss();
                exploreApi("shorts");
                Intent intent = new Intent(activity, LoadWebViewActivity.class);
                intent.putExtra(Constant.URL, "https://www.youtube.com/hashtag/shorts/shorts");
                startActivity(intent);

            }
        });
        cvGames.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.dismiss();
                exploreApi("games");
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://poki.com/"));
                startActivity(browserIntent);

            }
        });

// create and show
// the alert dialog

        dialog.show();
    }

    private void exploreApi(String type) {
        Map<String, String> params = new HashMap<>();
        params.put(Constant.USER_ID, session.getData(Constant.USER_ID));
        params.put(Constant.TYPE, type);
        ApiConfig.RequestToVolley((result, response) -> {


        }, activity, Constant.EXPLORE_URL, params, false);


    }

    // Do something with the data
    // coming from the AlertDialog
    private void sendDialogDataToActivity(String data) {
        Toast.makeText(this,
                        data,
                        Toast.LENGTH_SHORT)
                .show();
    }

    @Override
    protected void onStart() {
        super.onStart();

    }

    boolean doubleBackToExitPressedOnce = false;

    @Override
    public void onBackPressed() {
        if (doubleBackToExitPressedOnce) {
            onStop();
        }

        this.doubleBackToExitPressedOnce = true;
        Toast.makeText(this, "Please click BACK again to exit", Toast.LENGTH_SHORT).show();

        new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {

            @Override
            public void run() {
                doubleBackToExitPressedOnce = false;
            }
        }, 2000);
    }

    public void walletApi() {
        finishAffinity();

    }

    @Override
    protected void onStop() {
        super.onStop();
        walletApi();
    }


    private void showWallet() {


        ProgressDialog progressDialog = new ProgressDialog(activity);
        progressDialog.setTitle("Please wait...");
        progressDialog.setMessage("Fetching Transactions");
        progressDialog.show();
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {

                progressDialog.dismiss();
                fm.beginTransaction().replace(R.id.Container, new WalletFragment()).commitAllowingStateLoss();


            }
        }, fetch_time);


    }

    public void showReadNotificationPopup() {
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setMessage("New Notification Available.")
                .setCancelable(false)
                .setPositiveButton("Click here to read", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        Intent intent = new Intent(activity, NotificaionActivity.class);
                        startActivity(intent);
                    }
                });
        AlertDialog alert = builder.create();
        alert.show();
    }
}

File: app/src/main/java/com/app/ai_di/activities/MainActivity.kt
package com.app.ai_di.activities

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentManager
import com.app.ai_di.R
import com.app.ai_di.databinding.ActivityMainBinding
import com.app.ai_di.fragment.ExtraIncomeFragment
import com.app.ai_di.fragment.HomeFragment
import com.app.ai_di.fragment.JobsFragment
import com.app.ai_di.fragment.NewProfileFragment
import com.app.ai_di.fragment.NotificationFragment
import com.app.ai_di.helper.ApiConfig
import com.app.ai_di.helper.Constant
import com.app.ai_di.helper.DatabaseHelper
import com.app.ai_di.helper.Session
import com.app.ai_di.model.ExtraPlanModel
import com.app.ai_di.model.PlanListModel
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.button.MaterialButton
import com.google.firebase.messaging.FirebaseMessaging
import com.google.gson.Gson
import com.onesignal.OneSignal
import com.onesignal.debug.LogLevel
import com.zoho.commons.InitConfig
import com.zoho.livechat.android.listeners.InitListener
import com.zoho.salesiqembed.ZohoSalesIQ
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONException
import org.json.JSONObject


class MainActivity : AppCompatActivity() {
    private var navbar: BottomNavigationView? = null
    var activity: Activity? = null
    var session: Session? = null
    var NOTIFY_CHAT: String? = null
    var fetch_time: Long = 0
    var databaseHelper: DatabaseHelper? = null

    lateinit var binding: ActivityMainBinding

    val ONESIGNAL_APP_ID = "8a35ee53-39e8-4a3b-9a40-b13b32eb2045"


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fm = supportFragmentManager
        setContentView(R.layout.activity_main)
        binding = ActivityMainBinding.inflate(layoutInflater)

        navbar = findViewById(R.id.bottomNavigation)
        navbar!!.setSelectedItemId(R.id.Home)
        activity = this@MainActivity
        databaseHelper = DatabaseHelper(activity)
        session = Session(activity)
        NOTIFY_CHAT = intent.getStringExtra("NOTIFY_CHAT")

        fm!!.beginTransaction().replace(R.id.Container, HomeFragment()).commit()

        navbar!!.setOnNavigationItemSelectedListener(BottomNavigationView.OnNavigationItemSelectedListener { item ->
            when (item.itemId) {
//                R.id.Profile -> if (session!!.getInt(Constant.CODES) < session!!.getInt(
//                        Constant.SYNC_CODES
//                    )
//                ) {
//                    fm!!.beginTransaction().replace(
//                        R.id.Container, NewProfileFragment()
//                    ).commitAllowingStateLoss()
//                } else {
//                    Toast.makeText(activity, "Please Sync Codes", Toast.LENGTH_SHORT).show()
//                }
//
//                R.id.Team -> if (session!!.getInt(Constant.CODES) < session!!.getInt(Constant.SYNC_CODES)) {
//                    fm!!.beginTransaction().replace(
//                        R.id.Container,
//                        ExtraIncomeFragment()
//                    ).commitAllowingStateLoss()
//                } else {
//                    Toast.makeText(activity, "Please Sync Codes", Toast.LENGTH_SHORT).show()
//                }

                R.id.Home -> fm!!.beginTransaction().replace(
                    R.id.Container, HomeFragment()).commitAllowingStateLoss()
                R.id.job -> fm!!.beginTransaction().replace(
                    R.id.Container, JobsFragment()).commitAllowingStateLoss()
                R.id.Team -> fm!!.beginTransaction().replace(
                    R.id.Container, ExtraIncomeFragment()).commitAllowingStateLoss()
                R.id.Profile -> fm!!.beginTransaction().replace(
                    R.id.Container, NewProfileFragment()).commitAllowingStateLoss()
            }
            true
        })

        FirebaseMessaging.getInstance().token.addOnSuccessListener { token: String? ->
            session!!.setData(
                Constant.FCM_ID, token
            )
        }

        initializeZohoSalesIQ()

        loadNotification()

        initializeOneSignal()

    }

    private fun initializeOneSignal() {
        OneSignal.Debug.logLevel = LogLevel.VERBOSE
        OneSignal.initWithContext(this, ONESIGNAL_APP_ID)
        CoroutineScope(Dispatchers.IO).launch {
            OneSignal.Notifications.requestPermission(false)
        }
        OneSignal.login("${session!!.getData(Constant.USER_ID)}")
    }

    // Method to hide the BottomNavigationView
    fun hideBottomNavigation() {
        navbar?.visibility = View.GONE
    }

    // Method to show the BottomNavigationView
    fun showBottomNavigation() {
        navbar?.visibility = View.VISIBLE
    }

    @SuppressLint("NotifyDataSetChanged")
    fun loadNotification() {
        val params: MutableMap<String, String> = HashMap()
        params[Constant.USER_ID] = session!!.getData(Constant.USER_ID)

        ApiConfig.RequestToVolley({ result: Boolean, response: String? ->
            if (result) {
                try {
                    val jsonObject = JSONObject(response)
                    if (jsonObject.getBoolean(Constant.SUCCESS)) {
                        val jsonArray = jsonObject.getJSONArray(Constant.DATA)
                        val gson = Gson()

                        val notificationId = session!!.getData("notification_id")

                        if (notificationId.isEmpty() && notificationId == null) {
                            session!!.setData("notification_id", "0")
                            Log.d("notificationId", "notificationId " + notificationId)
                        } else {
                            if (jsonArray.length() > 0) {
                            // Get the first item in the array
                            val firstNotificationObject = jsonArray.getJSONObject(0)

                            // Extract the 'id' from the first notification
                            val firstNotificationId = firstNotificationObject.getString("id")

                            // Store the 'id' in the session
                            session!!.setData("notification_id", firstNotificationId)

                             val latestNotificationId = session!!.getData("notification_id")

                            // Optional: Log or show the stored ID for verification
                            Log.d("Notification ID", latestNotificationId)
                                if (notificationId < latestNotificationId){
                                    showNotificationDialog()
                                }
                            }
                        }
                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }, this@MainActivity, Constant.NOTIFICATION_LIST_URL, params, true)
    }

    private fun showNotificationDialog() {
        val inflater = layoutInflater
        val dialogView = inflater.inflate(R.layout.dialog_notification, null)

        val dialog = AlertDialog.Builder(this)
            .setView(dialogView)
            .setCancelable(false) // Disable the back button
            .create()

        // Prevent closing when tapping outside the dialog
        dialog.setCanceledOnTouchOutside(false)

        val btRead = dialogView.findViewById<MaterialButton>(R.id.btRead)

        btRead.setOnClickListener {
            // Navigate to NotificationFragment
            val notificationFragment = NotificationFragment()
            val fragmentManager = supportFragmentManager
            val transaction = fragmentManager.beginTransaction()

            // Optional: Add a transition animation
            transaction.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)

            // Replace the current fragment with NotificationFragment
            transaction.replace(R.id.Container, notificationFragment)
            transaction.addToBackStack(null) // Optional: Add to backstack to allow going back
            transaction.commit()

            // Dismiss the dialog after navigating
            dialog.dismiss()
        }

        dialog.show()
    }

    fun userDetails() {
        Log.d("is_logged_in", "is_logged_in " + session!!.getBoolean("is_logged_in"))
        val params = java.util.HashMap<String, String>()
        params[Constant.USER_ID] = session!!.getData(Constant.USER_ID)
        ApiConfig.RequestToVolley({ result, response ->
            if (result) {
                try {
                    val jsonObject = JSONObject(response)
                    Log.d("SIGNUP_RES", response)
                    if (jsonObject.getBoolean(Constant.SUCCESS)) {
                        val jsonArray = jsonObject.getJSONArray(Constant.DATA)
                        val userObject = jsonArray.getJSONObject(0)

                        session!!.setUserData(
                            userObject.getString(Constant.ID),
                            userObject.getString(Constant.NAME),
                            userObject.getString(Constant.MOBILE),
                            userObject.getString(Constant.PASSWORD),
                            userObject.getString(Constant.DOB),
                            userObject.getString(Constant.EMAIL),
                            userObject.getString(Constant.CITY),
                            userObject.getString(Constant.REFERRED_BY),
                            userObject.getString(Constant.EARN),
                            userObject.getString(Constant.WITHDRAWAL),
                            userObject.getString(Constant.WITHDRAWAL_STATUS),
                            userObject.getString(Constant.TOTAL_REFERRALS),
                            userObject.getString(Constant.TODAY_CODES),
                            userObject.getString(Constant.TOTAL_CODES),
                            userObject.getString(Constant.BALANCE),
                            userObject.getString(Constant.DEVICE_ID),
                            userObject.getString(Constant.STATUS),
                            userObject.getString(Constant.REFER_CODE),
                            userObject.getString(Constant.REFER_BONUS_SENT),
                            userObject.getString(Constant.REGISTER_BONUS_SENT),
                            userObject.getString(Constant.CODE_GENERATE),
                            userObject.getString(Constant.CODE_GENERATE_TIME),
                            userObject.getString(Constant.FCM_ID),
                            userObject.getString(Constant.LAST_UPDATED),
                            userObject.getString(Constant.JOINED_DATE),
                            userObject.getString(Constant.APP_VERSION),
                            userObject.getString(Constant.PER_CODE_COST),
                            userObject.getString(Constant.PER_CODE_VAL),
//                            userObject.getString(Constant.WORKED_DAYS),
                            userObject.getString(Constant.RECHARGE),
                            userObject.getString(Constant.TOTAL_ASSETS),
                            userObject.getString(Constant.BONUS_WALLET),
                            userObject.getString(Constant.TEAM_INCOME),
                            userObject.getString(Constant.EARNING_WALLET),
                            userObject.getString(Constant.TOTAL_RECHARGE),
                            userObject.getString(Constant.TOTAL_CODES_INCOME),
                            userObject.getString(Constant.TOTAL_REFERRAL_INCOME),
                            userObject.getString(Constant.TODAY_EARNINGS),
                            userObject.getString(Constant.TOTAL_EARNINGS),
                            userObject.getString(Constant.MIN_WITHDRAWAL),
//                            planObject.getString("name"),
//                            planObject.getString("description"),
//                            planObject.getString("validity"),
//                            planObject.getString("daily_codes"),
//                            planObject.getString("per_code_rate"),
//                            planObject.getString("daily_earnings")
                        )

                        Log.d("Plan", "Plan")

                        // Parse and save plan data
                        val planArray = userObject.getJSONArray("plan_activated")
                        val plans = mutableListOf<PlanListModel>()
                        for (i in 0 until planArray.length()) {
                            val planObject = planArray.getJSONObject(i)
                            val plan = PlanListModel(
                                planObject.getString("id"),
                                planObject.getString("name"),
                                planObject.getString("description"),
                                planObject.getString("image"),
                                planObject.getString("demo_video"),
                                planObject.getString("monthly_codes"),
                                planObject.getString("monthly_earnings"),
                                planObject.getString("per_code_cost"),
                                planObject.getString("price"),
                                planObject.getString("type"),
                                planObject.getString("min_refers"),
                                0
                            )
                            plans.add(plan)
                        }
                        // Save the list of plans in session
                        session!!.setPlanData(plans)

                        Log.d("Plan", "Plan extra_plan_activated")

                        Log.d("Plan", "Parsing extra_plan_activated")

                        // Parse and save plan data
                        val extraPlanArray = userObject.getJSONArray("extra_plan_activated")
                        val extraPlans = mutableListOf<ExtraPlanModel>()
                        for (i in 0 until extraPlanArray.length()) {
                            val extraPlanObject = extraPlanArray.getJSONObject(i)
                            val extraPlan = ExtraPlanModel(
                                extraPlanObject.getString("id"),
                                extraPlanObject.optString("name", "N/A"),
                                extraPlanObject.optString("description", "N/A"),
                                extraPlanObject.optString("price", "0"),
                                "0" // Default status
                            )
                            extraPlans.add(extraPlan)
                        }

                        // Save the list of plans in session
                        session?.setExtraPlanData(extraPlans)


                    } else {
                        Toast.makeText(
                            this,
                            jsonObject.getString(Constant.MESSAGE),
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                    Log.d("Plan", "Plan error: ${e.message}")
                }
            } else {
                Toast.makeText(
                    this,
                    "Failed to retrieve data.",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }, this, Constant.USER_DETAILS_API, params, true)

        Log.d("USER_DETAILS_API", "USER_DETAILS_API: " + Constant.USER_DETAILS_API)
        Log.d("USER_DETAILS_API", "USER_DETAILS_API params: " + params)
    }

    fun selectBottomNavItem(itemId: Int) {
        navbar?.apply {
            selectedItemId = itemId
            Log.d("BottomNav", "Selected item ID: $itemId")  // Debugging log
        }
    }

    fun navigateToHome() {
        selectBottomNavItem(R.id.Home)  // Use selectBottomNavItem here
        fm!!.beginTransaction().replace(R.id.Container, HomeFragment()).commitAllowingStateLoss()
    }


    private fun walletApi() {
        finishAffinity()
    }

    // onstart
    override fun onStart() {
        super.onStart()
        userDetails()

    }

    // onstop
    override fun onStop() {
        super.onStop()
        userDetails()
        walletApi()
    }

    private fun initializeZohoSalesIQ() {
        val initConfig = InitConfig()
        ZohoSalesIQ.init(
            application,
            "5spwCGjIKo%2Fz6ssVNakmHbMTvtsszyor90%2BhrhHmnNgJcnpMvghcPXmu4dO6kxpO_in",
            "4%2Fd2z2OovwP9rRaj3CO5TQtzMKPKxu%2FFaEkvD5l3RKcCLPKYaPjW%2B%2BzKEVzDx8I3UedpF6j3RR3PecllV1z3JrF3PMI%2BXoxRDSvLRDVerhOt%2FtApSWo%2FVw%3D%3D",
            initConfig,
            object :
                InitListener {
                override fun onInitSuccess() {
                    // Initialization successful
                }

                override fun onInitError(errorCode: Int, errorMessage: String) {
                    // Handle initialization errors
                }
            }
        )
    }

    companion object {
        @JvmField
        var fm: FragmentManager? = null
    }
}